Implement the AzureActiveDiretoryWebViewClient and related handlers

heidijinxujia opened the pull request:
Close #143 #160 #161

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/AzureActiveDirectoryWebViewClient.java:
```diff
@@ -0,0 +1,261 @@
+
+        if (isPkeyAuthUrl(formattedURL)) {
+            Logger.verbose(TAG, "WebView detected request for pkeyauth challenge.");
+            new PKeyAuthChallengeHandler(url, view, getRequest(), getCompletionCallback()).process();
```
My feedback for this line is quite similar to [this feedback about `NtlmChallengeHandler`](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/158#discussion_r200783221) -- mainly, that the construction and processing of the challenge should be the responsibility of separate objects

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/AzureActiveDirectoryWebViewClient.java:
```diff
@@ -0,0 +1,261 @@
+        return handleUrl(view, requestUrl.toString());
+    }
+
+    private boolean handleUrl(final WebView view, final String url) {
```
Aside from the critique about the challenge handlers, overall this method is pretty clean and readable 👍

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/EmbeddedWebViewAuthorizationStrategy.java:
```diff
@@ -69,7 +70,7 @@ public AuthorizationResult requestAuthorization(@NonNull final AuthorizationRequ
      * @throws ClientException
      */
     public EmbeddedWebViewAuthorizationStrategy(@NonNull final Activity activity,
```
We're increasing our coupling here between `EmbeddedWebViewAuthorizationStrategy` and Microsoft's services. I recall giving the feedback that the `EmbeddedWebViewAuthorizationStrategy` really ought to accept the `WebViewClient` as a parameter and this is exactly why.

If we're ever going to add support for other IdP's (and it is our goal to do this over a long enough timespan), this will have to get refactored out.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/OAuth2WebViewClient.java:
```diff
@@ -47,7 +48,7 @@
     private final IChallengeCompletionCallback mCompletionCallback;
     private final AuthorizationRequest mRequest;
-    private final Context mContext;
+    private final Activity mActivity;
```
This field appears to be used only in the creation of the `NtlmChallengeHandler` -- I suspect you can use [getContext()](https://developer.android.com/reference/android/view/View.html#getContext()) and avoid this entirely

heidijinxujia replied:
Actually, activity is required to construct the webview. And it is also used in the client cert request and other requests, like open link in browser and install brokers.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/OAuth2WebViewClient.java:
```diff
@@ -96,7 +97,7 @@ public void onReceivedHttpAuthRequest(WebView view, final HttpAuthHandler handle
         //TODO TelemetryEvent.setNTLM(true); after the Telemetry is finished in common.
         final NtlmChallengeHandler ntlmChallengeHandler
-                = new NtlmChallengeHandler(view, handler, host, realm, mContext, mCompletionCallback);
+                = new NtlmChallengeHandler(view, handler, host, realm, mActivity, mCompletionCallback);
```
Consider whether [`getContext()`](https://developer.android.com/reference/android/view/View.html#getContext()) will work for you purposes and consider removing the `Activity` here

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,5 +22,79 @@
+    private ClientCertRequest mClientCertRequest;
+    private Activity mActivity;
+
+    public ClientCertAuthChallengeHandler(@NonNull final ClientCertRequest request,
```
The `ClientCertRequest` is basically "the challenge", right? Unlike the other `ChallengeHandler` classes, this one appears to accept the Challenge and not construct it (which I like).

Consider the suggestion [from here](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/158#discussion_r200779395) --  you could declare this class using `IChallengeHandler<ClientCertRequest, Void>` and users of this class could conceivably swap out its implementation if they wanted to

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/PKeyAuthChallengeHandler.java:
```diff
@@ -22,21 +22,259 @@
+        mChallengeCallback = completionCallback;
+    }
+
+    public void process() {
```
Needs `@Override` but also I really don't see the value of this interface without accepting a parameter or return type.... 🤔

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,5 +22,79 @@
+    }
+
+    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
+    public void process() {
```
Needs `@Override`

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/PKeyAuthChallengeHandler.java:
```diff
@@ -22,21 +22,259 @@
+        return deviceCertProxy;
+    }
+
+    class PkeyAuthChallengeRequest {
```
This should be a top-level class. Not a child of its handler

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/PKeyAuthChallengeHandler.java:
```diff
@@ -22,21 +22,259 @@
+     * @param completionCallback
+     * @throws ClientException
+     */
+    public PKeyAuthChallengeHandler(@NonNull final String redirectUri,
```
This class really doesn't expose itself to be tested -- I think the `IChallengeHandler` interface needs some rethinking such that it provides some swappability/testability.

The `private` and `static` methods on this class do a TON of complex logic that we have no way to test.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/PKeyAuthChallengeHandler.java:
```diff
@@ -22,21 +22,259 @@
+     *
+     * @return Return Device challenge response
+     */
+    private PkeyAuthChallengeResponse getChallengeResponseFromUri() throws ClientException {
```
This is basically a factory method but because it's `private` we cannot test it -- how do you know this works??

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/PKeyAuthChallengeHandler.java:
```diff
@@ -22,21 +22,259 @@
+     *                    client must convey back>
+     * @return Return PKeyAuth challenge response
+     */
+    private PkeyAuthChallengeRequest getChallengeRequestFromUri(
```
This is also a factory method but is not testable

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/PKeyAuthChallengeHandler.java:
```diff
@@ -22,21 +22,259 @@
+        }
+    }
+
+    class PkeyAuthChallengeResponse {
```
Another top-level class

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/PKeyAuthChallengeHandler.java:
```diff
@@ -22,21 +22,259 @@
 
-    void processPkeyAuthChallenge() {
-        //TODO
+    enum RequestField {
```
_Maybe_ a top-level `enum` -- maybe just `Strings` (seems to only be used in this way)

heidijinxujia replied:
enum looks much cleaner to me. Not sure the benefits to change them into Strings constants.

piqiums commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/AzureActiveDirectoryWebViewClient.java:
```diff
@@ -0,0 +1,261 @@
+ */
+public class AzureActiveDirectoryWebViewClient extends OAuth2WebViewClient {
+    private static final String TAG = AzureActiveDirectoryWebViewClient.class.getSimpleName();
+    private static final String INSTALL_URL_KEY = "app_link";
```
this constant is also defined in "BasicWebViewClient", could we move it to AuthenticationConstants to keep only one definition?

heidijinxujia replied:
This is only used in this class. So I keep it inside the scope of this class.

piqiums replied:
I found it defined in 'BasicWebViewClient' on line 50  :  https://github.com/AzureAD/azure-activedirectory-library-for-android/blob/45db7dddf3ea20d5f558aca67203546924a1dd2b/adal/src/main/java/com/microsoft/aad/adal/BasicWebViewClient.java#L50

piqiums commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/AzureActiveDirectoryWebViewClient.java:
```diff
@@ -0,0 +1,261 @@
+            Logger.info(TAG, "Sending intent to cancel authentication activity");
+            Intent resultIntent = new Intent();
+            resultIntent.putExtra(AuthenticationConstants.Browser.RESPONSE_ERROR_CODE, parameters.get("error"));
+            resultIntent.putExtra(AuthenticationConstants.Browser.RESPONSE_ERROR_MESSAGE, parameters.get("error_description"));
```
please make these strings constants

piqiums commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/AzureActiveDirectoryWebViewClient.java:
```diff
@@ -0,0 +1,261 @@
+        // in the activity stack.
+        final int threadSleepForCallingActivity = 1000;
+        try {
+            Thread.sleep(threadSleepForCallingActivity);
```
Usually, we should avoid using 'Thread.sleep' since it might cause ANR. Could we try some other methods, like "Handler.postDelayed"?

piqiums commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/AzureActiveDirectoryWebViewClient.java:
```diff
@@ -0,0 +1,261 @@
+        resultIntent.putExtra(AuthenticationConstants.Browser.RESPONSE_ERROR_MESSAGE, errorMessage);
+
+        if (getRequest() != null) {
+            if (getRequest() instanceof AzureActiveDirectoryAuthorizationRequest) {
```
Please correct me if I was wrong, since MicrosoftStsAuthorizationRequest extends AzureActiveDirectoryAuthorizationRequest, these 2 case could be merged into one. Otherwise, the case of "if (getRequest() instanceof MicrosoftStsAuthorizationRequest)" won't be reached.

piqiums commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/ClientCertAuthChallengeHandler.java:
```diff
@@ -22,5 +22,79 @@
+        // When ADFS server sends null or empty issuers, we'll continue with cert prompt.
+        if (acceptableCertIssuers != null) {
+            for (Principal issuer : acceptableCertIssuers) {
+                if (issuer.getName().contains("CN=MS-Organization-Access")) {
```
please make this string a constant

piqiums commented on common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/PKeyAuthChallengeHandler.java:
```diff
@@ -22,21 +22,259 @@
+    public void process() {
+        mWebView.stopLoading();
+        mChallengeCallback.setPKeyAuthStatus(true);
+        new Thread(new Runnable() {
```
since you're posting a Runnable to the MessageQueue in the UI thread, you don't need to start a background thread to post the Runnable.

heidijinxujia commented:
@iambmelt  @piqiums  Thanks a lot for code reviews. I have addressed the comments. Have a look when you avail, as it still shows "requested changes". Thanks!

iambmelt commented:
Oops. Didn't see that there were conflicts in with the base branch -- let me know once you've resolved and I can reapprove 👍

heidijinxujia commented:
Due to the change of OAuth2ClientWebView PR. Need to resolve the conflict and keep a clean git commit history by creating a new PR #170
